src/data_device_manager/data_source.rs

```diff
@@ -0,0 +1,202 @@
+        source: &WlDataSource,
+        mime: Option<String>,
+    );
+
+    /// The client has requested the data for this source to be sent.
+    /// Send the data, then close the fd.
+    fn send_request(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        source: &WlDataSource,
+        mime: String,
+        fd: OwnedFd,
+    );
+
+    /// The data source is no longer valid
+    /// Cleanup & destroy this resource
+    fn cancelled(&mut self, conn: &Connection, qh: &QueueHandle<Self>, source: &WlDataSource);
+
+    /// A drop was performed
+    /// the data source will be used and should not be destroyed yet
+    fn dnd_dropped(&mut self, conn: &Connection, qh: &QueueHandle<Self>, source: &WlDataSource);
+
+    /// the dnd finished
+    /// the data source may be destroyed
+    fn dnd_finished(&mut self, conn: &Connection, qh: &QueueHandle<Self>, source: &WlDataSource);
+
+    /// an action was selected by the compositor
+    fn action(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        source: &WlDataSource,
+        action: DndAction,
+    );
+}
+
+impl<D, U> Dispatch<wl_data_source::WlDataSource, U, D> for DataDeviceManagerState
+where
+    D: Dispatch<wl_data_source::WlDataSource, U> + DataSourceHandler,
+    U: DataSourceDataExt,
+{
+    fn event(
+        state: &mut D,
+        source: &wl_data_source::WlDataSource,
+        event: <wl_data_source::WlDataSource as wayland_client::Proxy>::Event,
+        _data: &U,
+        conn: &wayland_client::Connection,
+        qh: &wayland_client::QueueHandle<D>,
+    ) {
+        match event {
+            wl_data_source::Event::Target { mime_type } => {
+                state.accept_mime(conn, qh, source, mime_type)
+            }
+            wl_data_source::Event::Send { mime_type, fd } => {
+                state.send_request(conn, qh, source, mime_type, fd);
+            }
+            wl_data_source::Event::Cancelled => {
+                state.cancelled(conn, qh, source);
+            }
+            wl_data_source::Event::DndDropPerformed => {
+                state.dnd_dropped(conn, qh, source);
+            }
+            wl_data_source::Event::DndFinished => {
+                state.dnd_finished(conn, qh, source);
+            }
+            wl_data_source::Event::Action { dnd_action } => match dnd_action {
+                WEnum::Value(dnd_action) => {
+                    state.action(conn, qh, source, dnd_action);
+                }
+                WEnum::Unknown(_) => {}
+            },
+            _ => unimplemented!(),
+        };
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct CopyPasteSource {
+    pub(crate) inner: WlDataSource,
+    pub(crate) serial: Option<u32>,
+}
+
+impl CopyPasteSource {
+    /// set the selection
+    /// internally tracks the serial for when unset selection may be called
+    pub fn set_selection(&self, device: &WlDataDevice, serial: u32) {
```

Comment:
I am not too sure what this is supposed to do from the documentation here.

Reply:
This method is really just a wrapper that calls set_selection. After looking at it again, I think it shouldn't hold onto the serial so I removed that part and adjusted the docs.